ast/ast.go

```diff
@@ -188,60 +234,140 @@ func (FunctionType) monotype() {}
 
 type NamedType struct {
 	BaseNode
-	ID *Identifier
+	ID *Identifier `json:"id"`
 }
 
 func (NamedType) Type() string {
 	return "NamedType"
 }
+func (nt *NamedType) Copy() Node {
+	if nt == nil {
+		return nt
+	}
+	nc := new(NamedType)
+	*nc = *nt
+	nc.BaseNode = nt.BaseNode.Copy()
+
+	nc.ID = nt.ID.Copy().(*Identifier)
+	return nc
+}
 
 type TvarType struct {
 	BaseNode
-	ID *Identifier
+	ID *Identifier `json:"id"`
 }
 
+func (c *TvarType) Copy() Node {
+	if c == nil {
+		return c
+	}
+	nc := new(TvarType)
+	*nc = *c
+	nc.BaseNode = c.BaseNode.Copy()
+
+	nc.ID = c.ID.Copy().(*Identifier)
+	return nc
+}
 func (TvarType) Type() string {
 	return "TvarType"
 }
 
 type ArrayType struct {
 	BaseNode
-	ElementType MonoType
+	ElementType MonoType `json:"elementType"`
 }
 
 func (ArrayType) Type() string {
 	return "ArrayType"
 }
+func (c *ArrayType) Copy() Node {
+	if c == nil {
+		return c
+	}
+	nc := new(ArrayType)
+	*nc = *c
+	nc.BaseNode = c.BaseNode.Copy()
+
+	//nc.ElementType = c.ElementType.Copy().(*ArrayType)
+	nc.ElementType = reflect.New(reflect.ValueOf(c.ElementType).Elem().Type()).Interface().(*ArrayType)
```

Comment:
Similar question here. Why the need for the `reflect` package? Does `Copy()` not work?